### ISSUE TITLE
Interface suggestions for new wigner-related calculations

### ISSUE THREAD

**Issue by zatkins2** [mreineck/ducc#26]
Hi Martin, 

I'm excited (via @xzackli) to use your in-development `coupling_matrix_spin0and2` methods! The ability to calculate coupling matrices for multiple spectra for roughly constant cost (independent of the number of spectra) will really enable some computations that are now prohibitive (since we currently calculate the same wigner symbols over and over for each spectrum separately).

To make it even more useful, could I suggest an additional interface? As it is now, all possible "flavors" of coupling matrix (00, 02, 20, ++, --, in that order) are calculated in `coupling_matrix_spin0and2`. Sometimes, not all of these "flavors" are necessary, e.g., we may want to skip calculating the last "--" flavor because it is negligible at our scales of interest. It would be nice to be able to tell the function which flavors we would like to compute (assuming the required spin-weighted spectra are passed), to save unnecessary computation. If it is no extra cost to compute all flavors though, then feel free to ignore me!

A separate request (that is lower priority for me personally :)) would be to be able to pass `l1min, l1max` to `wigner3j_int`, or even possibly an explicit list of `l1` values, rather than having the function calculate all possible values. But again, if this is not feasibly or doesn't save any time, also feel free to ignore me!

Many thanks, very much looking forward to using this functionality!

Zach

**Comment by mreineck** [comment]
Hi Zach,

great to hear that this might be useful!

Concerning the selection of coupling matrix flavors: as far as I can see at the moment (I've been working on other stuff recently, so this is all a little rusty :), there is not much performance to be gained by leaving out individual parts, with a few exceptions:
- you only need 00 (but for that we already have a separate routine)
- you only need ++ and -- (I don't have a special case for that, but the gain wouldn't be very big anyway)

That's for `coupling_matrix_spin0and2`. For the `_pure` variant things are more complex, but there also is only a very small subset of components where computing time can be saved.

There is however another aspect to this: if there is concern that the matrices become very big and you want to save memory (as opposed to time) by omitting some flavors, that can certainly be done, but I need some time for the adjustment.
(However if that is a concern, we could always save a nice factor of 2 by just returning just an upper triangular matrix and having the caller apply the `(2*l+1)` factors...)

Specifying an `l1` subset to `wigner3j_int` unfortunately does not save any time, since all possible coefficients must be computed anyway to get the proper normalization.

**Comment by zatkins2** [comment]
Good point, yes, memory could be an issue. If it is the case that the runtime is ~constant for large numbers of spectra, then I think being able to specify the exact flavor you want would be useful, even if it means ~2 or 3x the total computation. 

But I don't want to hold up getting what you have available in `ducc`; the current interface on your implementation would still be a massive improvement, so I am definitely eager to make use of it as soon as possible :)

(Another motivating example is that in practice it is often the case that the 02 spectra are equal to the 20 spectra)

**Comment by mreineck** [comment]
If you can already make use of the interface as it currently is, then I think it might be best to use it in this form for now. That will give us valuable information and allow us to come up with an even better interface in the next iteration.

Please let me know it I can help with integration in any way!

**Comment by zatkins2** [comment]
ah I'm not sure I can, at least, if I install `0.33`, I don't see the functions available. would I need to download the current state of the repo and install from that, rather than from pypi?

**Comment by mreineck** [comment]
Sorry, my bad: since these functions still have experimental interfaces, they are accessible as `ducc0.misc.experimental.coupling_matrix_spin0and2` etc.

**Comment by mreineck** [comment]
But they should be part of the official 0.33 release.

**Comment by zatkins2** [comment]
Got it, yep I see them, thanks! I will play around and let you know my experience (but I do expect I will want to be able to select the flavors)

**Comment by zatkins2** [comment]
I just played around a bit. It is really impressive! I observe that there is constant scaling with `nspec` up to a point, and then it becomes linear. The point when it switches to linear grows with `lmax`, since that allows the 3j computations to dominate. As a quick benchmark, for `lmax=10800`, with `nspec=1`, I find it takes ~3 min on 10 cores, and with `nspec=15`, I find it takes ~ 10 min. 

I do think it would be great if I could select the flavor. I know for the immediate application I have in mind, I will only need the 00, 02, and ++ flavors (so 40% speedup right away).

Another great option, like you mentioned, would be to not apply the 2l+1 factors. I will mostly be interested in building the symmetric coupling matrices, and could apply the 2l+1 factors later. Under the hood, I'm not sure if you're already doing this, but if not, this would allow you to compute only the upper triangle and then just copy the result to the lower triangle at the end (a 50% speedup).

Finally, if it could support single precision, that would be fantastic (another 50% speedup). 

To be clear, I am making all these suggestions because this implementation is absolutely fantastic!

**Comment by zatkins2** [comment]
And yes besides speedup, eg single precision will be important for SO for memory. The combinatorics of the covariance matrix will require a huge number of couplings to be calculated

**Comment by mreineck** [comment]
Just curious: did you use a binary wheel or did you build from source, using `pip install --no-binary ducc0`? If the former, you can get another nice speedup by compiling yourself.

> I do think it would be great if I could select the flavor. I know for the immediate application I have in mind, I will only need the 00, 02, and ++ flavors (so 40% speedup right away).

I will certainly think about this!

> Another great option, like you mentioned, would be to not apply the 2l+1 factors. I will mostly be interested in building the symmetric coupling matrices, and could apply the 2l+1 factors later. Under the hood, I'm not sure if you're already doing this,

I do :) So there will not be much speedup from this ...

I hesitated using the "half-matrix" format because it makes the structure of the output array more obscure; but halving memory is just too big of an advantage...

> Finally, if it could support single precision, that would be fantastic (another 50% speedup).

A large part of the calculations probably need to stay in double precision to avoid error accumulation. Input and output can certainly be in single precision however!

**Comment by zatkins2** [comment]
> Just curious: did you use a binary wheel or did you build from source, using pip install --no-binary ducc0? If the former, you can get another nice speedup by compiling yourself.

Yep I did!

> I will certainly think about this!

Great, thanks! Would also help with memory.

> I do :) So there will not be much speedup from this ...

Ok sure. It would still be nice to let the user decide about the 2l+1 factor I think. Practically speaking anyway, there are many more matrices computed for the covariance (which don’t have that factor) than for the spectra (which do). 

> A large part of the calculations probably need to stay in double precision to avoid error accumulation. Input and output can certainly be in single precision however!!

Awesome! No worries on the computation but yes for memory management it would be a great option.

**Comment by xzackli** [comment]
Thank you again @mreineck, I just want to reiterate Zach's comments that your work here has significant analysis implications for ACT and SO covariances, which grinds through a tremendous number of these coupling matrices.

**Comment by mreineck** [comment]
I have added a new method `coupling_matrix_spin0and2_tri`, which allows selecting the accuracy of the output and returns triangular matrices (see the docstring for memory layout).

You'll have to install this one directly from the repo, it's on branch `mcm_work`.

Concerning selectable flavors: to get good performance I fear that I'll have to write a custom implementation for every combination of provided input spectra and required output matrices.
It' not a big deal however (I expect around 10-15 minutes of work for adding a new combination).

@zatkins2, did I understand correctly that in your first test case you want to provide three spectra instead of four , assuming that the 02 spectrum is the same as the 20 one, and that you only need the first, second and fourth matrix instead of the full output?

**Comment by mreineck** [comment]
One more thing I should perhaps add: given the limited memory bandwidth of a compute node, it might well happen that there may be an optimal number of spectra to compute simultaneously, instead of the idealized "the more, the better" expectation. This will depend on `lmax`, the number of threads and the underlying hardware (especially cache sizes). To get the absolute maximum performance out of the algorithm, some experimentation will be required, but for large data sets this should be worth the effort.

**Comment by zatkins2** [comment]
> I have added a new method coupling_matrix_spin0and2_tri, which allows selecting the accuracy of the output and returns triangular matrices (see the docstring for memory layout).
>
> You'll have to install this one directly from the repo, it's on branch mcm_work.

Very cool! I will try it out. As far as the triangular layout goes, do you have a (parallelized!) function that can turn it into the triangular half of a 2d array? That I think would be necessary. Alternatively, Sigurd has one [here](https://github.com/simonsobs/pixell/blob/a5e44701c4f63106bc105f48c14d99f8fe233ca9/pixell/curvedsky.py#L705), which would work where `iainfo` and `oainfo` correspond to the triangular and rectangular layout, but only if you were to change your triangular layout to match what he adheres to [here](https://github.com/simonsobs/pixell/blob/a5e44701c4f63106bc105f48c14d99f8fe233ca9/pixell/curvedsky.py#L394) (it looks like they differ).

> Concerning selectable flavors: to get good performance I fear that I'll have to write a custom implementation for every combination of provided input spectra and required output matrices.
It' not a big deal however (I expect around 10-15 minutes of work for adding a new combination).
>
> @zatkins2, did I understand correctly that in your first test case you want to provide three spectra instead of four , assuming that the 02 spectrum is the same as the 20 one, and that you only need the first, second and fourth matrix instead of the full output?

Understood and that's correct. I think having the current version available, this proposed version available, and two more would cover every case that would actually happen:
- consume three spectra (assuming 02 is the same as 20), and produce the first, second, fourth, and fifth matrix
- consume all four spectra, and produce all but the fifth matrix

Another random thought: if we knew that the spectra themselves were the same, could that help optimize things? E.g. in the current implementation, if all four spectra were the same, would there be some way of calculating the five matrices faster or using less memory? That situation will be common.

**Comment by mreineck** [comment]
I think the two memory layouts are actually the same, if `m` corresponds to `l1` and `l` to `l2` ... but I admit that this is not immediately obvious from the formulas. So you should be able to use `pixell` routines on the arrays.

From the custom routines, I have implemented the proposed one, under the provisional name `coupling_matrix_spin0and2_tri_zach1` :-)
The others will probably be ready later today.

> Another random thought: if we knew that the spectra themselves were the same, could that help optimize things? E.g. in the current implementation, if all four spectra were the same, would there be some way of calculating the five matrices faster or using less memory? That situation will be common.

Yes, that will most likely help things. But again, this will be something that has to be hardwired into the sources in some way, `if/else` branches at runtime will most likely slow things down.

**Comment by mreineck** [comment]
The subtle point about Sigurds formula
`mstart = stride*(m*(2*lmax+1-m)//2)`
is that it is actually equivalent to
`mstart = stride*((m*(2*lmax+1-m))//2)`
and not equivalent to
`mstart = stride*(m*((2*lmax+1-m)//2))`
as I read it first.
With that misunderstanding out of the way, I convinced myself that the storage schemes are in fact identical.

**Comment by zatkins2** [comment]
> I convinced myself that the storage schemes are in fact identical.

Great.

> From the custom routines, I have implemented the proposed one, under the provisional name coupling_matrix_spin0and2_tri_zach1 :-)
> The others will probably be ready later today.

Amazing, thank you so much Martin! 

> Yes, that will most likely help things. But again, this will be something that has to be hardwired into the sources in some way

Got it, I will leave it as food-for-your-thought.

**Comment by mreineck** [comment]
The `_zach2` and `_zach3` variants are now also in the code, but very much untested. Have fun!

**Comment by zatkins2** [comment]
Ok! Will work off the `mcm_work` branch for now and let you know if any comments/issues. Thanks again.

**Comment by mreineck** [comment]
How about this fully generic interface for `coupling_matrix_spin0and2_tri`
Instead of the "full" version
 `coupling_matrix_spin0and2_tri(spec, lmax, nthreads=nthreads, singleprec=True)`
one could write
 `coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,2,3), mat_index=(0,1,2,3,4), nthreads=nthreads, singleprec=True)`
`spec_index` and `mat_index` describe where the individual spectra/matrices are located, and if the index values are negative, the respective spectrum is not provided / the matrix is not requested.

The `zach1`, `zach2` and `zach3` versions would then look like this:
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,1,2), mat_index=(0,1,-1,-1,2), nthreads=nthreads, singleprec=True)`
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,1,2), mat_index=(0,1,-1,2,3), nthreads=nthreads, singleprec=True)`
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,2,3), mat_index=(0,1,2,3,-1), nthreads=nthreads, singleprec=True)`

Using this syntax, it would also be easy to express that all spectra are equal (`spec_index=(0,0,0,0)`).
I think I can mirror this on the C++ side using templating, which could make the sources significantly shorter.
However, I'd still need to explicitly instantiate all the variants of the function we need, i.e. it won't be possible to access all imaginable combinations even after this change.

Do you think this would be a useful change?

**Comment by mreineck** [comment]
Sorry ... I just was impatient and pushed the change :-)
I'm pretty sure this is a step in the right direction. There is now only a single function for all the triangular coupling matrices, called `coupling_matrix_spin0and2_tri`, which supports all the variants discussed above, plus pure spin 0, plus (for experimenting) a version that computes all matrices, but assumes that all spectra are equal. Please have a look at the docstring and let me know if anything is unclear!

Out of curiosity: do you expect to use the `spin0and2_pure` function as well? I had expected that this would be required for accurate calculations.

**Comment by mreineck** [comment]
I pushed another change that accelerates `coupling_matrix_spin0and2_tri` when only a small number of spectral sets is provided, so don't be surprised if timings for 1-4 sets (pure spin0) and 1-2 sets (all other cases) look strange compared to other configurations.

Overall I'm now fairly happy with the current state. The only aspect I don't like very much is the really long compile time...

**Comment by zatkins2** [comment]
Sorry for not replying sooner. Your consolidation looks nice to me!

Would the `zach1` case not correspond to this:
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,1,2), mat_index=(0,1,-1,2,-1), nthreads=nthreads, singleprec=True)`
(notice the difference in the `mat_index`)?

For clarity, I think there are 4 discussed cases, including your original one, that would look like this in the new interface. Lmk if you agree:
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,1,2), mat_index=(0,1,-1,2,-1), nthreads=nthreads, singleprec=True)`
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,1,2), mat_index=(0,1,-1,2,3), nthreads=nthreads, singleprec=True)`
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,2,3), mat_index=(0,1,2,3,-1), nthreads=nthreads, singleprec=True)`
`coupling_matrix_spin0and2_tri(spec, lmax, spec_index=(0,1,2,3), mat_index=(0,1,2,3,4), nthreads=nthreads, singleprec=True)`

**Comment by mreineck** [comment]
You are right, and this should now be fixed!
(I'm now wondering if I got the original `zach1` version right ...)

**Comment by mreineck** [comment]
Let me come back to the single vs. double precision for a moment ...

It seems that `pspy` does all coupling-matrix related computations in `real(8)` aka `double`, and `NaMaster` computes the 3j symbols in `double` and the rest in `flouble`, which is presumably `float` in most applications.

Do you think we are safe in all "normal" situations if we use the `NaMaster` approach? The only thing I'm a bit worried about is the accumulation in the loops over `el3` (or `l1` in `NaMaster`).

**Comment by zatkins2** [comment]
Which approach did you go with initially? I would agree with you that I would be wary of accumulating large sums in a single-precision buffer.

On my end, I think the most important thing is that the output buffer would be in single precision (if that's what was requested), and that results that have been fully accumulated in double precision are copied "on-the-fly" into the output buffer. What I mean by "on-the-fly" is if you are computing `nspec*4` couplings, that you don't have `nspec*4` intermediate double-precision buffers, but rather 1 double-precision buffer that gets reused `nspec*4` times or something like that; otherwise, there isn't a memory savings for requesting single-precision since there must be enough memory available to hold the intermediate `nspec*4` double-precision buffer.

Also, random question from earlier: in the `tri` method, does it no longer apply the `2l+1` factor to the output, since that `2l+1` factor is not symmetric, but to "unravel" it would most naturally assume symmetry?

**Comment by mreineck** [comment]
> Which approach did you go with initially? I would agree with you that I would be wary of accumulating large sums in a single-precision buffer.

I'm currently accumulating into double precision numbers and, once the accumulation is finished, I'm copying to the output array, which can be single or double precision. At any one time, only a handful (<100) accumulations take place per thread, so additional memory consumption is not a concern. 

As a rough benchmark, I'm now running full `spec0and2` calculations (single-precision output) with `lmax=10800` and `nspecs=30` in 30min on my laptop (8 cores), and the process requires 36GB.

> Also, random question from earlier: in the tri method, does it no longer apply the 2l+1 factor, since that is not symmetric?

That's correct.

**Comment by zatkins2** [comment]
Sounds great.

With the 0.33.0 version of this interface, I got that `nspec=15` and the same `lmax` was calculated in 10min on 10 cores on a Princeton cluster node (and, demonstrating the awesome utility of your implementation, `nspec=1` was calculated in 3min, so very sub-linear scaling!)

**Comment by mreineck** [comment]
I hope that the timings on the Princeton cluster will improve somewhat with the current version - not dramatically, but it should be noticeable. The most important change will be the reduced memory consumption (factor 1/4 when using the `tri` method and single precision matrices).

**Comment by mreineck** [comment]
I have now removed the older `spin0` and `spin0and2` methods that returned square matrices; maintaining them in parallel to the triangular ones would add quite a lot of work for practically no benefit.
The `_pure` method still exists, and I'm happy to look into further optimizations there once I understand the typical usage scenarios.

**Comment by mreineck** [comment]
Since the original suggestions are now implemented and (I hope) working properly, here a few follow-up questions:
- what should I do about `coupling_matrix_spin0and2_pure`? Is this actually going to be used? If not, I'd probably comment it out for the time being. If it is used, should I make it return triangular matrices as well, for a more homogeneous interface (some of the matrices will be transposed however)?
- Is there a need for an equivalent to `pspy`'s `calc_covariance` functions?

@zatkins2, @xzackli any thoughts?

**Comment by zatkins2** [comment]
Re the pure methods, I can't comment. I personally won't be using them, but they may be of use to others in ACT/SO working on dedicated B-mode analyses.  I can ask some of my colleagues this week and report back to you (or have them comment). Or yes, Zack might have an opinion.

Can you link the `calc_covariance` method you are thinking of? I don't know it off the top of my head.

**Comment by zatkins2** [comment]
(I just asked Susanna Azzoni, an SO BB group person, what the group does now; she says they use `namaster`. I will try to evangelize `ducc` to them, maybe they have a desire for faster code!)

**Comment by mreineck** [comment]
> Can you link the calc_covariance method you are thinking of? I don't know it off the top of my head.

I was thinking about the code here: https://github.com/simonsobs/pspy/blob/master/pspy/cov_fortran/cov_fortran.f90
This is called mainly from `pspy/so_cov.py` and some `pspy` tutorials.
The code structure is very similar to that of the the mode-coupling functions, so it should be straightforward to optimize it in the same way.

**Comment by mreineck** [comment]
> (I just asked Susanna Azzoni, an SO BB group person, what the group does now; she says they use namaster. I will try to evangelize ducc to them, maybe they have a desire for faster code!)

If I read it correctly, `namaster` recently switched to `ducc` for its SHTs, so getting the mode-coupling matrices from `ducc` as well may not be such a big change. Let's see :-)

**Comment by zatkins2** [comment]
I see. I would say that's less urgent (I only use the low-level calls to coupling matrices, not these wrappers)

**Comment by mreineck** [comment]
I have now merged the `mcm_work` branch into `ducc0`. We can start a new branch when more adjustments are necessary.

**Comment by mreineck** [comment]
I released `ducc0 0.34` today, which contains the new functionality.

**Comment by mreineck** [comment]
I'll close this for now, since there is nothing concrete to do until we have some solid user experience. Please reopen (or open a new issue) if you have any other suggestions!

**Comment by mreineck** [comment]
Just curious: has this been successfully used in the meantime? Any lessons learned that suggest changing/enhancing the interface?

**Comment by xzackli** [comment]
@zatkins2 is leading the charge here, I think he can say more!

**Comment by mreineck** [comment]
I'd still be very interested to hear if this code is actively used anywhere ... if only to justify to people at MPA why I spent time on it :-)

**Comment by xzackli** [comment]
Happy New Year! @zatkins2 now has published [2412.07068](https://arxiv.org/abs/2412.07068) which relied on this code in development (equations in 4.1).

**Comment by xzackli** [comment]
Not sure where the code ended up...

**Comment by mreineck** [comment]
Doesn't seem as if the new code was used for the paper; that's perfectly fine from my side, I just like to know when it *is* used somwhere :-) Also I already have other potential use cases, for example within Euclid, so this definitely wasn't wasted time.
@xzackli, do you think it would make sense to push for inclusion in, say, `NaMaster` and/or `pspy`? Usually I don't like to nag people, but at the same time I expect that users of these packages might be quite happy about this kind of speedups...

**Comment by zatkins2** [comment]
Hi both, sorry for any confusion and missing your question over the summer @mreineck. The short answer is/was, it has not been used in the meantime (but not for any issue with your code or its interface!). Rather, the ACT DR6 results were too far along (essentially complete) by then, so there wasn't really much of an opportunity to pull/test the changes and make use of their benefits.

However, this should (will?) be the baseline code for coupling matrix calculations for the SO LAT power spectrum analysis (which hasn't begun). I.e., we should replace the equivalent fortran code in `pspy` with calls to `ducc`, and we are in the right window of time now to be doing that. (I only add a ? because I'm not fully in charge of `pspy`, but I don't see why this wouldn't happen). I will open a `pspy` issue tomorrow and work on the PR.

Sorry again for the haphazard timing, I definitely do not take your work for granted :) To summarize, I think it will be heavily used in the SO LAT analysis!

**Comment by zatkins2** [comment]
Hi Martin, 

Hope all is well in the new year. Sorry again for the slow-going on this, but I just wanted to report that these codes are now the core part of the SO LAT power spectrum pipeline. This is true both for the computation of mode-coupling matrices (for estimating two-point power spectra) and the symmetric coupling matrices for the covariance matrix. For the covariance, your code is critical for making the calculation possible: since SO will ultimately have ~5 times the number of maps as ACT, the covariance is O(625x) larger, which would have been prohibitively slow with the old code. (I still have not used the "pure" functions since I don't really work on the SATs still).

You can follow code on the `iso` branches of [`PSpipe`](https://github.com/simonsobs/PSpipe/tree/iso/project/SO/pISO/python) and [`pspy`](https://github.com/simonsobs/pspy/tree/iso). The simple function that wraps ducc is just [this](https://github.com/simonsobs/pspy/blob/9bfc43ff1c4fcde7b6da62fa7d06c4e088ca8e78/pspy/so_mcm.py#L18). It recurs in many scripts which are much more optimized than before thanks to its speed.

I'm finding a few features of the current implementation in ducc that would be great to discuss more. There are really 3:

1. 
For the covariance, the mapping from the four spins of the four fields to the "spin flavor" of the resulting coupling matrix is non-trivial. As a result, the covariance computation doesn't map so cleanly onto the existing `ducc` API in which the 00, 02, 20, and 22 window spectra must be "packaged together."

It would actually be best to not use `calc_coupling_spin0and2()` as an inspiration, and instead to do something even more granular: just let the user supply a list of window spectra each with a specified spin type. For example, there may be 10 "00" windows, 4 "02" windows, and 2 "22" windows. If a window spectra is 00, it can only produce a 00 matrix. If a window is 02, it can only produce a 02 matrix (which is the same as a 20 matrix for the same spectrum). If a window is 22, it can produce one or both of the ++ and -- matrices.

I'm not sure how the underlying C works, but this could perhaps be implemented in just 3 separate methods (one each for 00, 02, and 22 window spectra). One reason I say this is it looks like a `ducc` call to get the `5N` 00, 02, 20, ++, and -- matrices is greater than 5x longer than to get the `N` 00 matrices. Perhaps separate functions would be just as fast, if not faster? But that's just speculation, the more important thing is the API: allowing arbitrary numbers of window spectra with arbitrary spins.

Otherwise, now we need to make repeated calls to `ducc` simply because we don't need the same number of the different kinds of matrices, which wastes computation.

2.
My earlier suggestion for returning the matrix in "triangular" format was totally off-base. It would be ideal if the matrices were just returned in square format (as symmetric matrices). You can see in the `pspy` function, we always need to make the triangular-to-square conversion anyway, but it's less efficient and wastes memory (we need to hold both the triangular and square versions in memory at the same time). Sorry about that. 

3.
I'm wondering if it would be possible to implement the "Toeplitz" approximation (see https://arxiv.org/abs/2010.14344 and [here](https://github.com/simonsobs/PSpipe/blob/d474d80a5462999a74929b3e1533892023299ec5/project/ACT_DR6/python/get_covariance_blocks.py#L75)). In practice, this means basically computing couplings exactly to a lower `lmax` than desired, and then doing simple copy-paste operations to go out to `lmax`. For the covariance, this approximation is basically accurate, and is much faster than the exact coupling all the way to `lmax`. Because there are so many terms in the covariance, it would really help out there.

Let me know what you think! With the code in-place now, I really am in a position to rapidly incorporate any of the above changes if you agree with them (and have the time and desire to work on them!). Sorry again that that was not the case before.

**Comment by mreineck** [comment]
Hi Zach,

very good to hear that the integration is making progress!

Let's look at the individual points and see what can be done:

Concerning 1), I actually think that your suggestion is already implemented behind the scenes. Depending on the content of `spec_index` and `mat_index`, completely different low-level functions are called to do the actual work, and each of them is optimized to do only the necessary work. (They have the same name in fact, but since they are massively templated depending on the contents of `spec_index` and `mat_index`, they are different pieces of machine code.)

> One reason I say this is it looks like a ducc call to get the `5N` 00, 02, 20, ++, and -- matrices is greater than 5x longer than to get the `N` 00 matrices.

That is totally expected; the 00 matrices only require Wigner 3j symbols with `m1=m2=m3=0`, which are more than twice as fast to compute than ones with nonzero `m`. Also, half of these symbols are zero (if I remember correctly), which makes the dot products faster as well.

So I think if you can sort the window spectra into homogeneous groups with the same `spec_index` and `mat_index` respectively, you'll get optimal performance.
Lumping different `spec_index` and `mat_index` into a single call could be allowed in principle, but internally I'd still have to disentangle the various "families" and compute them with separate functions, because performance will suffer otherwise.

Perhaps I'm also completely missing the point, in which case it would be great to have a concrete example!


2) Producing full symmetric matrices should be doable; I'll check if I can implement this without too much copy-pasted code!


3) Assuming that we switch to square output matrices, couldn't we simply achieve this by calling the `ducc` functions with a lower `lmax` and providing a partial view of the final result array as `res`? Then you can Toeplitz-extend it afterwards...

**Comment by mreineck** [comment]
Since you were mentioning `pspy`: do you think we can get a discussion going on https://github.com/simonsobs/pspy/issues/97?
If my assumption there is wrong, then I need to adjust the `ducc` code :-)

**Comment by mreineck** [comment]
I had a closer look at the Toeplitz paper, and according to Figure 3 the part of the array that needs to be computed is quite a bit more complicated than just reducing `lmax`. It should be possible to just compute the required part of the array, but it may be good to first verify how much of a performance advantage this approach still has (e.g. by comparing the time for the full, accurate computation to the time for the Toeplitz filling).

**Comment by zatkins2** [comment]
Thanks for your thoughts!

> I actually think that your suggestion is already implemented behind the scenes. Depending on the content of spec_index and mat_index, completely different low-level functions are called to do the actual work, and each of them is optimized to do only the necessary work. (They have the same name in fact, but since they are massively templated depending on the contents of spec_index and mat_index, they are different pieces of machine code.)
> 
> So I think if you can sort the window spectra into homogeneous groups with the same spec_index and mat_index respectively, you'll get optimal performance.
Lumping different spec_index and mat_index into a single call could be allowed in principle, but internally I'd still have to disentangle the various "families" and compute them with separate functions, because performance will suffer otherwise.
> 
> Perhaps I'm also completely missing the point, in which case it would be great to have a concrete example!

No, you understand well. To be clear, those low-level functions are not exposed right now right? The issue is just then that the  current python function enforces the "lumping into a single call." Would it be easy to expose the low-level functions for each of the different cases? That sounds ideal! I.e., exposing something like `calc_coupling_matrix_spin00`, `calc_coupling_matrix_spin02`, and `calc_coupling_matrix_spin22`. The only complication would be for the last one, whether the user wants the output `++`, `--`, or both. These would accept `spec` arrays of shape `(nspec, nl)`. 

(Then we also no longer need to worry about which combinations of `spec_index` and `mat_index` are useful, it's just up to the user to manage. Though you could keep them as well if you want, perhaps others will find them useful still).

> Producing full symmetric matrices should be doable; I'll check if I can implement this without too much copy-pasted code!

Amazing, that would also be ideal!

> I had a closer look at the Toeplitz paper, and according to Figure 3 the part of the array that needs to be computed is quite a bit more complicated than just reducing lmax. It should be possible to just compute the required part of the array, but it may be good to first verify how much of a performance advantage this approach still has (e.g. by comparing the time for the full, accurate computation to the time for the Toeplitz filling).

Yes. I do think the speedup is quite worth it (though it's not an apples-to-apples comparison with the speedup that would happen in `ducc`, since I am using the slow `pspy` methods for the exact coupling):

On 8 cores, the following exact coupling:
```python
lmax = 8500 

wcl = np.random.randn(2*lmax)

l_toep, l_band, l_exact = lmax, lmax, lmax

mcm = np.zeros((lmax, lmax))
so_mcm.mcm_fortran.calc_coupling_spin0(wcl, l_exact, l_band, l_toep, mcm.T)

if l_toep < lmax:
    mcm = so_mcm.format_toepliz_fortran2(mcm, l_toep, l_exact, lmax)

so_mcm.mcm_fortran.fill_upper(mcm.T)
```
ran in 183 seconds. Replacing the third line with `l_exact, l_band, l_toep = 800, 2000, 2750` (the standard Toeplitz parameters used in `PSpipe`) reduces it to 18 seconds. So I think the "filling" operation is quite fast compared to the calculation, at least in `pspy`.

That being said, yes, the picture in Fig. 3 is a bit more complicated, but I think some simplifications of it would be just fine for the covariance matrix. The path forward depends on two things:

1. It's necessary that regardless of what happens at low `lmax`, the diagonal of the matrix is computed to the full `lmax`. In your algorithms, is this something that could be fast (i.e., faster than calculating all elements to the full `lmax`)? 

2. For the picture in Fig. 3, yes, we would also need to calculate full specific columns of the matrix (but only a subset). This seems less plausible to calculate quickly than the full diagonal, but I don't know. 

Even if only 1. is possible, I think that would still be fine, we don't need to follow the picture in the paper. Instead, I would propose a simpler picture that should work just fine for the covariance: calculate the exact square coupling up to some lower `lmax`, then fill the "band" covered by that "square" along the diagonal and the even-farther-off-diagonals could just be left 0. The only thing beyond a lower `lmax` that would be needed is the exact, true-diagonal up to the full `lmax` still.

**Comment by zatkins2** [comment]
For completeness, in the speed test using the Toeplitz approximation, the "filling" steps are about 0.5 seconds, the other 17.5 seconds is the `pspy` coupling. So yes, it is definitely a worthy speedup even with `ducc` couplings. (As above, we might want to do a different filling operation, but it would be similar).

**Comment by mreineck** [comment]
Thanks, I think I now understand the idea with the more flexible interface better. Basically you say that you only have a "handful" of spectra for which you want to compute the coupling matrices, and to use `coupling_matrix_spin0and2_tri` "properly", you'd have to call it more than once in different `spec_index` and `mat_index` modes, which of course means redundant computation of 3j symbols. Alternatively you could call it in "full" mode, providing some dummy spectra and taking some dummy matrices in the result; that would actually run at almost perfect performance, but admittedly it is a waste of memory.
I *think* I can do something there, but the interface will be "interesting". Basically, `spec_index` and `mat_index` would go away (which I think is nice), `spec` would have shape `(nspec, lmax_spec+1)`, `res` would have shape `(nmat, lmax+1, lmax+1)`, and there would be an additional input array or tuple of length `nspec` describing what to do with every individual spectrum in the input, like:
- 0: spectrum is of type 00, append a 00 coupling matrix to the output
- 1: spectrum is of type 02 or 20, append a 02 or 20 (which is the same) coupling matrix to the output
- 2: spectrum is of type 22, append a ++ coupling matrix to the output
- 3: spectrum is of type 22, append a -- coupling matrix to the output
- 4: spectrum is of type 22, append a ++ *and* a -- coupling matrix to the output

`nmat` would be uniquely defined by the transform types of the individual spectra.

Does this sound roughly correct?

BTW, the value of `nspec` is pretty important for the performance of the function. If at all possible,, getting this above 10 or so would be great, perhaps even if that means changing the structure of the calling code a bit. Avoiding redundant 3j computations is *really* beneficial.

**Comment by mreineck** [comment]
> For completeness, in the speed test using the Toeplitz approximation, the "filling" steps are about 0.5 seconds, the other 17.5 seconds is the `pspy` coupling. 

That is still with the traditional coupling code, right?
I'm asking because we may get a factor 10 or more with the new coupling code, and while the full computation is most likely still more expensive than the filling step, the advantage of never having to worry about approximation quality may be compelling.
Still, let's get to that point first and then see.

**Comment by zatkins2** [comment]
> Thanks, I think I now understand the idea with the more flexible interface better. Basically you say that you only have a "handful" of spectra for which you want to compute the coupling matrices, and to use coupling_matrix_spin0and2_tri "properly", you'd have to call it more than once in different spec_index and mat_index modes, which of course means redundant computation of 3j symbols. Alternatively you could call it in "full" mode, providing some dummy spectra and taking some dummy matrices in the result; that would actually run at almost perfect performance, but admittedly it is a waste of memory.

That's right.

>I think I can do something there, but the interface will be "interesting". Basically, spec_index and mat_index would go away (which I think is nice), spec would have shape (nspec, lmax_spec+1), res would have shape (nmat, lmax+1, lmax+1), and there would be an additional input array or tuple of length nspec describing what to do with every individual spectrum in the input, like:
>
>    0: spectrum is of type 00, append a 00 coupling matrix to the output
>    1: spectrum is of type 02 or 20, append a 02 or 20 (which is the same) coupling matrix to the output
>    2: spectrum is of type 22, append a ++ coupling matrix to the output
>    3: spectrum is of type 22, append a -- coupling matrix to the output
>    4: spectrum is of type 22, append a ++ and a -- coupling matrix to the output
>
>nmat would be uniquely defined by the transform types of the individual spectra.
>
>Does this sound roughly correct?

Yes, exactly! I find that interface much simpler and more flexible than the existing one. 

> BTW, the value of nspec is pretty important for the performance of the function. If at all possible,, getting this above 10 or so would be great, perhaps even if that means changing the structure of the calling code a bit. Avoiding redundant 3j computations is really beneficial.

Yes, it is not necessarily true that we "only have a "handful" of spectra," it may be many, but the point is just that the number of 00 matrices != the number of 02 matrices != the number of ++ matrices etc., hence the need for repeated function calls (or yeah, extra memory like you said). On the point of `nspec`, I find on the tiger cluster at Princeton that the optimal number is around 32. After that, the time per matrix starts to actually increase. 

> That is still with the traditional coupling code, right?
I'm asking because we may get a factor 10 or more with the new coupling code, and while the full computation is most likely still more expensive than the filling step, the advantage of never having to worry about approximation quality may be compelling.
Still, let's get to that point first and then see.

Sorry, I reported the timings in a confusing way. The relevant quantity there is that the "filling" operation took 0.5 seconds. For the same number of cores and `lmax`, the call to `ducc` is about 18s for `nspec=1` and 66s for `nspec=32`. So as long as our "filling" operation is the same or similar, it is subdominant currently, so in principle the Toeplitz approximation is worth it.

For example, for `lmax=4000`, `ducc` drops to ~3s per call. So long as the diagonal from `4000` to `8000` can be computed for not too much extra cost (I have no idea), it's possible a "Toeplitz" ducc matrix for these parameters may take as little as 3.5 seconds including the "filling." Yes, if `nspec` became large enough, I guess the "filling" time would grow faster than the `ducc` time. But even for `nspec=32`, that's 16 seconds of "filling," and currently about 66 seconds of exact `ducc` coupling -- so still worth it.

But, point taken. The improvement will definitely be less (probably a factor of ~3) in `ducc`, not a factor of ~10 in `pspy`. But the ASO covariance will have millions of matrices ;), so even a few seconds saved adds up to node-hours :( 

> the advantage of never having to worry about approximation quality may be compelling.

Yes, to be clear, we would not use this for the estimation of the two-point spectra which we would want to be exact, only for the estimation of their analytic covariance matrix. That's just because the number of matrices for the covariance is orders of magnitude larger than for the spectra. For the spectra, the exact matrices are plenty fast as-is.

The covariance calculation involves many approximations already. In fact, the use of these coupling matrices itself is already an approximation. So just for the covariance, I wouldn't have this worry (I mean, all the approximations need to be good enough, but I think this additional one would be too).

**Comment by mreineck** [comment]
There is a working (but not yet fully optimized) version on branch `coupling_matrices`. Please give it a try!
(Note that the matrices are now fully populated, but not yet multiplied by `2l+1`.)

**Comment by zatkins2** [comment]
Amazing! Yes, let's please keep it that way -- _don't_ multiply by the rows by `2l+1`! Or, if you want to add this as a feature in `ducc`, make it optional :) But we'll still need to wrap this function (minimally) regardless, and multiplying by `2l+1` is very easy. 

And yes, Toeplitz is not a high priority right now, we can return to it. In short, I am basically just curious whether you think calculating only the `l1 = l2` matrix elements could be done "much faster" than all the matrix elements.

**Comment by mreineck** [comment]
Sure, no problem to leave out the `2l+1` factor! I suspect that the routine will never be part of "general purpose" `ducc`, so we have a bit of flexibility :-)

It's no problem to just fill certain parts of the arrays, and I intend to start with this as soon as we are happy with the full-array version.
You just need to be aware that the vectorized Wigner 3j symbols are only fast if we can compute them for ranges
`l1=[a; a+veclen]` with `l2` being constant and larger than all `l1`. In other words, if we only compute the diagonal, performance per matrix element will drop by a factor of 4 or 8.
However, since we also have dense areas in the regions required by the Toeplitz approximation, I think we should be OK, because vectorization will work there.

**Comment by zatkins2** [comment]
Okie doke. Will test and let you know, but I expect it is basically ready to ship.

> You just need to be aware that the vectorized Wigner 3j symbols are only fast if we can compute them for ranges
l1=[a; a+veclen] with l2 being constant and larger than all l1. In other words, if we only compute the diagonal, performance per matrix element will drop by a factor of 4 or 8.

Just for my understanding, how does the second sentence follow from the first? It seems like the first sentence implies that no elements of the diagonal beyond the dense part of the coupling matrix can be vectorized; instead, each one follows the other in a not-vectorized for-loop. Is the vectorization speedup per-element a factor of 4 to 8 on average?

**Comment by mreineck** [comment]
I meant to say that if we have vector instructions, whenever we compute the Wigner symbols for given `l1, l2, m1, m2`, we can get additional symbol vectors for different `l1,l2` more or less for free, provided
- they have the same `m1, m2`
- their `l3` range has the same length as for the original `l1,l2` pair.
- their `l1,l2` are "reasonably similar" to the original.

If the CPU has AVX, the resulting speedup is close to 4, for AVX512 it can be close to 8.
Assuming we only want 3j symbols for `l1==l2`, we can't satisfy the second condition, so we lose the performance advantage.

**Comment by mreineck** [comment]
I thought about an easy way to compute just the required part of the matrices for the Toeplitz approxiimation.
The most obvious would be to have
- a loop for the diagonal
- a loop for the vertical line at `l_toeplitz`
- a double loop for everything below `l_exact`
- a double loop for the remaining part

However that's a lot of effort. Instead I'd like to loop over the entire `l1,l2` range as before, and for every concrete combination of `l1,l2` I just check whether it falls into the required region. If not, I skip this loop iteration.
So everything boils down to a selection function `neededForToeplitz(l1, l2, l_exact, l_toeplitz, dl_band)`, which returns `True` or `False`. This has the advantage of allowing easy future adjustments to the required area.
What do you think?

**Comment by mreineck** [comment]
I just implemented this; it was extremely straightforward.
To use partial matrix computation, just specify values for `l_exact`, `l_toeplitz` and `dl_band`!

**Comment by mreineck** [comment]
To clarify, partial matrix computation will populate a few more matrix elements than strictly necessary, This is expected, but harmless, and you can simply overwrite them with the Toeplitz-approximated values.
I don't recommend to use these extra values ... they are correct, but if you ever switch to a machine with different vector length, your results will silently change, which is not a good thing.

**Comment by zatkins2** [comment]
Wow! Ok. Sorry to be clear: it is not doing any “filling” of the non-calculated values? That still has to happen outside ducc? (Besides the extra ones you mentioned). 

Do you see a large speedup?

**Comment by mreineck** [comment]
That's correct, I'm not filling in all the "non-accurate" values since I don't have a good understanding how that works. I expected that you had standalone functions for exactly this purpose, but I admit I didn't look closely at that part of your implementation.
Speedup factors are similar to what's reported in the paper, which is what I would expect (i.e. slightly better than the ratio of accurately calculated values to total values).

**Comment by mreineck** [comment]
You can look at what the function does by filling the result array with zeros, then calling the `ducc` method with the proper Toeplitz parameters, and finally doing a `plt.imshow(res[0]==0.)` or similar.

**Comment by zatkins2** [comment]
Gotcha, thanks. That’s fine, yes, I just wanted to make sure I needed to call the filling function that exists.

**Comment by mreineck** [comment]
I'll try to implement automatic "Toeplitz-filling" some time next week; I think I understand the algorithm sufficiently now.

**Comment by zatkins2** [comment]
Everything looks very nice on the new branch! Just a couple very minor comments:

1. The `_tri` method seems to be slightly faster for the same "request," but perhaps the difference is just because the `_new` method has to fill the upper triangle? For example, with `nthreads=32`, `lmax=8500`, `singleprec=True`, `_new` with `spec.shape = (20, 21600)` (ignore that `spec` is longer than it needs to be, I don't think it matters) and `optype = [0]*5 + [1]*10 + [4]*5` takes ~35 seconds, but `_tri` with `spec.shape = (5, 4, 21600)`, `spec_index=(0, 1, 2, 3)`, and `mat_index=(0, 1, 2, 3, 4)` takes ~30 seconds. This slight decrease in speed with `_new` is completely fine! And with Toeplitz it is way faster than this anyway. But noting it for you just in case it points you to some low-hanging-fruit under the hood.
2. Toeplitz works great. I have a filling function, but yes, if you don't mind implementing it too, then it saves one step that needs to be called in a `pspy` wrapper. Either way is totally fine by me, no need to do it yourself, just let me know.

**Comment by mreineck** [comment]
Yes, I think I know the reason for the slowdown; it's probably caused by the additional flexibility of the function. Since every spectrum now could potentially be used for generating a "--" matrix, I have to load its value not only for index `l`, but also for index `l+1` in a critical inner loop, and this single machine instruction makes a big difference.
There may be a way around this, but I didn't have time yet to give it a try.

**Comment by zatkins2** [comment]
No worries, this shouldn't matter much ultimately. I will go ahead and write my wrapper around the branch now, but let me know if you make more updates and if/when you deploy to the main branch (and what the new function name is)!

Thanks for all this!

**Comment by mreineck** [comment]
I have now committed an updated code that should do the Toeplitz filling. However I'm not totally sure yet that it does the same thing as current `pspy`, since at least for the "02" case the spectra seem to have to fulfill some specific criteria to avoid negative entries on the matrix diagonal. Please give this a very thorough testing!

**Comment by zatkins2** [comment]
Ok, nice. Can you elaborate what the criteria are that you have in mind? You are probably correct that such criteria exist, and `pspy` does not handle anything special right now (i.e., it assumes any such criteria are met).

**Comment by mreineck** [comment]
Oh, I don't really know the criteria. But if I just take completely random spectra (as I did in my test script), I get negative values on the matrix diagonals, and the Toeplitz filler tries to take square roots from them, which leads to a lot of NaN values.

**Comment by zatkins2** [comment]
I see. Actually I don’t think this is unique to 02. For the others, yes, the 3j symbol may be squared, but the spectrum may be a “cross”-spectrum between different fields and can also take on negative values. I agree I’m not sure it’s so easy to know in advance if the diagonal will have negative values, but I think for most “sensible” combinations of masks, the diagonal will be positive.

Given that the Toeplitz call is already so much faster than exact, I’d advocate for this solution: if the diagonal (including and beyond l_exact) has any negative values, repeat that matrix with an exact computation. I feel pretty confident that this should be rare. 

To be clear, I don’t think you should implement this, I think it’s best done in python. E.g., a user would want to “stack” all such offending matrices in a single follow-up ducc call. So I would implement this in a wrapper. What do you think?

**Comment by zatkins2** [comment]
I tested your filling function and mine and they returned identical results. I also reviewed your code and it looks good.

**Comment by mreineck** [comment]
Thanks for having a look at this! I think I'll leave the code as it is for a week or so and then give the function a better name and merge it to the main branch.
If you can think of any improvement, please let me know!

**Comment by zatkins2** [comment]
Btw, here is the wrapper that handles (among other things) rerunning the bad Toeplitz matrices exactly by testing the output diagonals: https://github.com/simonsobs/pspy/blob/befe0ffe29354201e6a96f9ec4640c263e7dd854/pspy/so_mcm.py#L16

One thought actually: what do you think about being able to specify the lmax of the rows and columns separately? Again, just thinking for the covariance, in the medium-term we will want to calculate the covariance of two spectra with different lmaxs (e.g., 3000 and 8000), so it would be good to save time vs. calculating an 8000x8000 matrix and then slicing out more than half of it.

**Comment by mreineck** [comment]
Thanks! The recomputation step looks good!

Concerning the different `lmax`: I think this should not be hard to do; we should probably just agree on whether `l1max` or `l2max` will always be the larger one, to simplify the structure of the function a little bit.

**Comment by zatkins2** [comment]
Sure, let’s make the rows the larger one?

**Comment by mreineck** [comment]
Just to make absolutely sure (since this code mixes Fortran- and C-Code): that is the first of the two matrix indices? ;-)

**Comment by mreineck** [comment]
Ah, nevermind, I just realized that this can be trivially adjusted if necessary. I'll just go ahead and implement something in the next days.

**Comment by zatkins2** [comment]
Ok, sure, no rush at all. I was going to say that rows = 0th axis of C ordered, 2d numpy array. But I just picked that it should be longer randomly

**Comment by mreineck** [comment]
I just realized that this is a bit more complicated than it looked at first - nothing dramatic, but there will be many small changes that might introduce subtle errors.
If it's OK I'd prefer to stick to the current functionality for the time being, and re-evaluate additional features after we are certain that everything we have now is rock solid.

**Comment by mreineck** [comment]
BTW, CPU time difference for full matrix and inequal `l1max` and `l2max` should be pretty small, especially in Toeplitz mode. Memory saving would of couse still be nice.

**Comment by zatkins2** [comment]
Ok, good to know the under-the-hood speedup is minimal/not worth the effort! Yes, the memory reduction ultimately becomes a practical speedup, because it means you can stuff more couplings into one call. Given that, I'd propose instead a hopefully simple solution:

When the output buffer is allocated, make it the requested "inequal lmax1 and lmax2" shape, and (sort of like your solution for checking whether an element is a Toeplitz element), only write into it if the element is inside the bounds. But the _calculation_ itself can still go out to the largest lmax along both axes. 

Would that be simpler?

**Comment by mreineck** [comment]
That one should work, yes!

Good point about the speedup caused by calling the function with more couplings per call!

**Comment by mreineck** [comment]
Hmm, I'm wondering whether our version of Toeplitz filling can work on "partial" matrices ...

There are lines like
```
            mat(l1,l2) = T(row_toep[l2-l1]*diag[l1]*diag[l2]);
```
That can only work if the full diagonal is known, and we don't have that available in the `l1max!=l2max` case.

**Comment by zatkins2** [comment]
I think it would be OK, but yes it's a good point it would require some rejiggering of how (as I understand it) you have organized the high-level coupling methods. Basically, either: you would need to Toeplitz-fill the output buffer at the same time that calculations are written into the output buffer (rather than doing the entire Toeplitz-fill at the end of the coupling method), or at least the `row_exact`, `row_toep`, and `diag` arrays would need to be filled at calculation-time and then fed into the Toeplitz-fill function at the end of the coupling method  (rather than filling them as part of the Toeplitz-fill function itself).

My hunch is the second one would be easier perhaps, but not sure. Either way, this is not urgent!

**Comment by mreineck** [comment]
I just pushed a new version, which hopefully has all the features. Please note that the interface has changed!

`lmax` and `singleprec` are gone. `res` is now a mandatory array, from which `l1max`, `l2max` and `singleprec` are inferred.
The last two dimensions of `res` can now be different, and it doesn't matter which one is larger.

I'm pretty sure that there will be broken corner cases, e.g. when `l_toeplitz` lies between `l1max` and `l2max` or similar. Please let me know whenever you encounter incorrect results or crashes, even when the inputs may seem silly!

**Comment by zatkins2** [comment]
Cool! I won’t have a ton of time to test this in the next couple weeks so will assume edge cases don’t affect me 😅😅 (still doing “squares” for now). But will take a look when I can.

**Comment by zatkins2** [comment]
Actually -- if it's OK with you, is it possible for the previous commit (before the interface change) to be put on a branch, so we can still download it easily? Perhaps like `coupling_matrices_beta` or something, since it is more stable? I've been using that version now and it's amazing. Basically, I don't have the bandwidth to check for bugs in the new one right now, but do need to distribute my code to others in the meantime

**Comment by mreineck** [comment]
Sure! I just pushed `coupling_matrices_beta`; please check whether everything works as expected.

**Comment by mreineck** [comment]
I'm going to release 0.41 in the near future.  Unless there are further ideas for the interface, the current version will go in as it is.

**Comment by zatkins2** [comment]
Ok! We've been stuck on `coupling_matrices_beta`. If that works for you, could that be the version that goes into the release?

**Comment by mreineck** [comment]
I'd much prefer not to go back to this version. Other people have already switched to the new implementation, and maintaining both seems a waste (both versions are absolute compile time hogs). I can keep `coupling_matrices_beta` in the repo for some time of course, but it probably won't get updates.

**Comment by zatkins2** [comment]
Ah sorry, ok, yes, please leave the `beta` branch for now and I will work to update our code to the released veresion. Out of curiosity, do you know who else has been using these methods (if you can share)?

**Comment by mreineck** [comment]
I'm sorry for causing confusion ... by "further ideas for the interface" above I meant incremental small changes or additions, not the rather big reversion :-)

Some PhD students and postdocs at MPA are using the code, and Elina Sihvola uses the code for Euclid work (https://arxiv.org/abs/2506.18676).

**Comment by zatkins2** [comment]
Right ok, err sorry yes just to be clear: you mean the `coupling_matrices` branch (with differing `lmax`s along the two sides) will be released, but you'll keep the `coupling_matrices_beta` branch in the repo for now?

Sorry I haven't been able to test the differing-`lmax`s version. Do you know if anyone encountered the broken corner cases you suggested?

**Comment by mreineck** [comment]
Exactly. The `coupling_matrices` branch has been merged, and `coupling_matrices_beta` remains as a tag.

So far I had no one do thorough tests with different `l1max` and `l2max` unfortunately. But we should be able to sort out any problems quickly if they are found.

**Comment by mreineck** [comment]
Just FYI: there has been interesting progress for coupling matrix computation: have a look at https://arxiv.org/abs/2602.15605!
Their code is very compact and efficient, and, when comparing apples to apples, faster than the current `ducc` implementation. It lacks the "process multiple spectra at once" feature though.
I'll look into this code during the next days and try to add multi-spectra support and explicit vectorization. Let's see where the new performance limit is!

**Comment by zatkins2** [comment]
Yes I have seen this paper and spoken with Georgia (@gkiddier22)! She is also working on joining SO too which will be great. Yes, the most important feature is the "process multiple spectra," but very nice to see it is faster than the current implementation on a per-coupling basis (@merryduparc also found this when the paper went out, but I should have poked you about it too, apologies). Very interested to see what you find!

**Comment by mreineck** [comment]
For me the nicest aspect is that the implementation is so small and can be auto-vectorized by the compiler to a large extent. I have already found a few tweaks to make it even faster (https://github.com/gkiddier22/threej_cosmo/pull/2), but haven't started with a prototype inside `ducc` yet.

**Comment by zatkins2** [comment]
Cool! Happy to integrate whatever interface you two decide is best

**Comment by zatkins2** [comment]
By the way, this reminds me to ask -- to what extent can `ducc` be installed on a GPU? Or, it would be ideal if this method could be run on a GPU (as Georgia's paper shows)! 

Right now the bottleneck for me is actually not the coupling matrix computation, but rather the summation of 10^4  x 10^4 arrays (of which the coupling matrix is one "layer"). This would naturally be faster on a GPU, but potentially not if the couplings need to be computed on a CPU first and then moved to the GPU (though that is speculation on my part).

**Comment by mreineck** [comment]
`ducc` is pure CPU, sorry. That might change at some point in the future, but only if I find a good, vendor-neutral way to express all the algorithms.

That said, I'm curious what makes the operation you mention (which should be `O(lmax**2)`) slower than the coupling matrix stuff, which is `O(lmax**3)`. Can you point me to an example?

**Comment by zatkins2** [comment]
Sure, I'd love to hear any suggestions you have! Sorry this is a lot of info here, but trying to be as concise as possible.

If you wanted, we could have a call as well, that may be more efficient than a github comments conversation :) I would be very happy to do that. If you're interested please email me! (zatkins@sas.upenn.edu)

##

The main script is this https://github.com/simonsobs/PSpipe/blob/zach_piso/project/SO/pISO/python/get_ana_cov_blocks.py. There is quite a bit of book-keeping that precedes this script so that it can run the fewest computations possible (basically, by reusing couplings as much as possible rather than naively recomputing them for each combination of four maps and polarizations. Also, some other well-motivated approximations such as treating masks that are very similar as being the same, etc).

For a given MPI process with a fixed amount of memory, we compute a set of covariance blocks that won't exceed the memory of the process -- this sets the maximum number of couplings we can compute at once (say, 32 couplings or so, which in my testing roughly saturates the speedup from the "multiple spectra at once" feature of `ducc`). The point of the script is that we may be able to compute hundreds of covariance blocks (a "set") using only these 32 couplings. The couplings for a given set of covariance blocks are computed [here](https://github.com/simonsobs/PSpipe/blob/a474fcfbc137135cbc698d66b2370dff5168dc85/project/SO/pISO/python/get_ana_cov_blocks.py#L188).

Then, we need to compute each covariance block in the set, in [this loop](https://github.com/simonsobs/PSpipe/blob/a474fcfbc137135cbc698d66b2370dff5168dc85/project/SO/pISO/python/get_ana_cov_blocks.py#L208). Each block consists of all 81 (3^4) polarization combinations, which are then looped over [here](https://github.com/simonsobs/PSpipe/blob/a474fcfbc137135cbc698d66b2370dff5168dc85/project/SO/pISO/python/get_ana_cov_blocks.py#L245). Each polarization combination then consists of adding several "layers" of `(8000, 8000)` single-precision arrays* [here](https://github.com/simonsobs/PSpipe/blob/a474fcfbc137135cbc698d66b2370dff5168dc85/project/SO/pISO/python/get_ana_cov_blocks.py#L307). Each layer is the simple product of a coupling matrix, and two 2d arrays that are each the "outer sum"  of a 1d `(8000,)` array, as in [this function](https://github.com/simonsobs/PSpipe/blob/a474fcfbc137135cbc698d66b2370dff5168dc85/project/SO/pISO/python/get_ana_cov_blocks.py#L134). Again, the point of the book-keeping in this script is essentially to call this line as few times as possible, while grouping as many couplings as possible into the call to `ducc`. 

(*You can see, I've also added into the script the option to "pre-bin" the arrays by a given factor `cov_prebin_size` to try to cut-down these arrays from `(8000, 8000)` to something smaller, but this is not always a good approximation.)

It turns out that on average there are ~430 terms to add to each block (there is a very wide scatter though). In other words, 430 calls to line 307 of the script. But the average number of couplings per block is about 0.6, with nearly all blocks requiring 2 or fewer couplings. I don't remember precisely, but on 4 cores, the cost of adding one "term" is like 50ms on average (so, ~20s total per block for all ~400 "terms" on average), whereas the cost of a single coupling (given that we compute 32 at a time, use Toeplitz etc) is like 1s on average (so ~1s per block).

TLDR: basically, (a) using Toeplitz reduces the `lmax` for the coupling vs. the array addition, and (b) the algorithm we use to be efficient in our book-keeping reduces the number of couplings needed way below the number of 2d arrays, so the prefactor is much less.

I'm quite bummed about how slow it is to add these arrays, and have found there is likely some inefficient allocation of memory and/or memory-bottlenecked step when the results of line 307 are added together, because as processes complete, the remaining processes speed up greatly, and in testing the function of line 307 on its own, it runs in about 17ms on 4 cores.

**Comment by mreineck** [comment]
First numbers are in: combining the new TT algorithm with the `ducc` optimizations, I can do a single `lmax=10000` coupling matrix in 4.8s on an 8-core laptop. The official released `ducc` takes 15.75s for the same task.
With increasing number of simultaneous spectra the difference decreases of course ...

**Comment by mreineck** [comment]
Results for the analogous EE case: 23.78s vs. 79.36s. L2 error between implementations below 1e-14.
Yes, this is a nice algorithm!

Now for 02 and -- components ...

**Comment by zatkins2** [comment]
That is very impressive indeed! 

> The official released ducc

Did you compile it on your laptop or download from pip 😛 

> With increasing number of simultaneous spectra the difference decreases of course ...

But presumably, when this feature is merged with the new algorithm, the 3x speedup for TT and EE would remain for any number of spectra?

**Comment by mreineck** [comment]
I compiled all versions by hand, so the speedup is genuine :-)

> But presumably, when this feature is merged with the new algorithm, the 3x speedup for TT and EE would remain for any number of spectra?

No, because each spectrum adds a (to first order) constant `delta_t` to the overall time, which is the same for both implementations - so the speedup reaches 1 for very many spectra.

**Comment by zatkins2** [comment]
Is it the same constant for both algorithms?

And, assuming it is the same constant, isn't that only true for a _very many_ spectra? I.e., isn't the limit where the 3j computation still dominates more important (and in this limit, the 3x speedup of threej_cosmo vs. ducc would persist for more than 1 spectrum)?

**Comment by mreineck** [comment]
> Is it the same constant for both algorithms?

Pretty much. All that changes is how we compute the Wigners themselves.

> And, assuming it is the same constant, isn't that only true for a very many spectra? I.e., isn't the limit where the 3j computation still dominates more important (and in this limit, the 3x speedup of threej_cosmo vs. ducc would persist for more than 1 spectrum)?

It's somewhere in between ... see the runs below.
For your typical batch size of 30 you wouldn't really notice the difference.

```
martin@smalliron:~/codes/ducc$ python3 python/demos/mcm_demo.py 

Mode coupling matrix computation comparison
nspec=1, lmax=5000, nthreads=8

Spin 0 case:
ducc square time (single precision): 2.137097120285034s
ducc square new time (single precision): 0.6732096672058105s
1.859041310876706e-15
martin@smalliron:~/codes/ducc$ python3 python/demos/mcm_demo.py 

Mode coupling matrix computation comparison
nspec=10, lmax=5000, nthreads=8

Spin 0 case:
ducc square time (single precision): 5.858177900314331s
ducc square new time (single precision): 4.3987977504730225s
1.855431865776542e-15
martin@smalliron:~/codes/ducc$ python3 python/demos/mcm_demo.py 

Mode coupling matrix computation comparison
nspec=30, lmax=5000, nthreads=8

Spin 0 case:
ducc square time (single precision): 13.180861473083496s
ducc square new time (single precision): 11.703989505767822s
1.8599017409362937e-15
```

**Comment by zatkins2** [comment]
I see thanks! For 30 spectra, the `new` time is about 17.5x the 1 spectrum time, whereas for the `old` time it was only 6x. I guess the new algorithm is so fast that the Wigner part quickly becomes subdominant!

But good to see at least 17.5 is still a healthy amount less than 30...

**Comment by mreineck** [comment]
>  I guess the new algorithm is so fast that the Wigner part quickly becomes subdominant!

Yes! Applying the 3j symbols to each power spectrum still requires `O(lmax**3)` multiplications and additions ... there is not much we can do about that.

Still, please note that this is for TT, which is an extreme case. For the other types of coupling matrices, the actual Wigner computation part is way more dominant.

**Comment by mreineck** [comment]
Here are the numbers for EE:

```

Mode coupling matrix computation comparison
nspec=1, lmax=5000, nthreads=8

EE case:
ducc square time: 9.79170846939087s
ducc square new time: 2.9443142414093018s
2.4251573330529034e-15
martin@smalliron:~/codes/ducc$ python3 python/demos/mcm_demo.py 

Mode coupling matrix computation comparison
nspec=10, lmax=5000, nthreads=8

EE case:
ducc square time: 13.61014175415039s
ducc square new time: 6.038481950759888s
2.154861532330252e-15
martin@smalliron:~/codes/ducc$ python3 python/demos/mcm_demo.py 

Mode coupling matrix computation comparison
nspec=30, lmax=5000, nthreads=8

EE case:
ducc square time: 20.669606924057007s
ducc square new time: 12.408429861068726s
2.1521590141475487e-15
```

**Comment by mreineck** [comment]
All types of coupling matrices are now supported. You can check the new version on the `new_wigners` branch, using the function `ducc0.misc.experimental.coupling_matrix_rect_new()` (same interface as `coupling_matrix_rect`).

Please let me know if you notice anything unexpected!

**Comment by mreineck** [comment]
Timings for a single spectrum:

```
TT case:
ducc square time: 2.0040547847747803s
ducc square new time: 0.6045055389404297s
1.8581924545064092e-15

EE case:
ducc square time: 9.830324172973633s
ducc square new time: 2.8463332653045654s
2.426257701359066e-15

TE case:
ducc square time: 12.710071086883545s
ducc square new time: 3.145857334136963s
3.6205380301458225e-14

EB case:
ducc square time: 9.939424991607666s
ducc square new time: 2.497058391571045s
2.3757006974497746e-15
```

For 10 spectra:

```
TT case:
ducc square time: 5.258899927139282s
ducc square new time: 4.03308629989624s
1.862011713423697e-15

EE case:
ducc square time: 13.351032972335815s
ducc square new time: 5.777651071548462s
2.158984830496376e-15

TE case:
ducc square time: 16.34982657432556s
ducc square new time: 5.840753555297852s
3.667418132314406e-14

EB case:
ducc square time: 13.425783157348633s
ducc square new time: 5.360791921615601s
2.158674512730641e-15
```

So multiple spectra still help a lot for EE/TE/EB, and a bit for TT.

**Comment by mreineck** [comment]
I modernized the benchmark script and ran a few different scenarios, with the following result:

```
Mode coupling matrix computation comparison
lmax=5000, nthreads=8

('00',) case:
ducc square time: 2.0829885005950928s
ducc square new time: 0.62742018699646s
L2 error: 1.91176179776569e-15

('++',) case:
ducc square time: 9.080282211303711s
ducc square new time: 2.650601625442505s
L2 error: 3.3089496142554628e-15

('02',) case:
ducc square time: 11.967025995254517s
ducc square new time: 2.5177032947540283s
L2 error: 2.6534211595552507e-15

('--',) case:
ducc square time: 9.09711742401123s
ducc square new time: 2.578315019607544s
L2 error: 3.0302128518774482e-15

('++/--',) case:
ducc square time: 9.363731861114502s
ducc square new time: 5.2058210372924805s
L2 error: 3.2077930603370896e-15

('00', '02', '++', '--', '++/--') case:
ducc square time: 17.026405334472656s
ducc square new time: 9.015365362167358s
L2 error: 2.7539117745326503e-15

('00', '00', '00', '00', '00', '00', '00', '00', '00', '00') case:
ducc square time: 5.515681266784668s
ducc square new time: 4.369277477264404s
L2 error: 1.934176623413761e-15

('00', '02', '++/--', '00', '02', '++/--', '00', '02', '++/--', '00', '02', '++/--', '00', '02', '++/--') case:
ducc square time: 26.421955108642578s
ducc square new time: 17.178869247436523s
L2 error: 2.639375087046633e-15
```

Overall, I think the code is now ready to merge to the `ducc0` branch, but I propose to leave it under the name `coupling_matrix_rect_new()` for now, until it has got some real-world testing.

**Comment by zatkins2** [comment]
Looks good! Will try the new branch and the other suggestions from our conversation and report back. I expect no issues

**Comment by mreineck** [comment]
I actually just merged, so you can do your tests directly on `ducc0` :-)

**Comment by zatkins2** [comment]
But still with `_new` in the function name?

**Comment by mreineck** [comment]
Yes!
